### PR TITLE
fix: include CSRF token in schedule provider selection

### LINKED ISSUE #3428: [Bug]: Schedule provider last-name selection fails with CSRF 403
## Description

The provider last-name search at the top-right of the Schedule page displays matching providers, but selecting a result fails with an error instead of switching the schedule to that provider.

The generated provider-selection form submits a POST request to `/provider/providercontrol` without the required CSRF token, so the request is rejected with HTTP 403.

## Steps to Reproduce

1. Log in with a local synthetic development account.
2. Open the Schedule page.
3. Enter `test` in the provider last-name search at the top-right and submit the search.
4. Select a provider from the results popup.

## Expected Result

The selected provider schedule opens successfully.

## Actual Result

The provider selection fails and the server returns HTTP 403.

```text
POST /carlos/provider/providercontrol HTTP/1.1 403
CSRF violation: Required Token is missing from the Request
```

## Module/Area

Schedule / provider last-name search

## Technical Notes

- Confirmed on deployed `develop` commit `b79ba7c6c6c5be5ce8f6db1988987cc66f17f324`.
- `src/main/webapp/WEB-INF/jsp/provider/receptionistfindprovider.jsp` creates the POST form dynamically in `selectProvider()` but only adds the provider/preference fields; it does not add the CSRF token.
- The initial search popup is opened with a GET. The failure occurs during the follow-up POST after a search result is selected.
- Related but distinct from #3255, which only tracks the 10-character limit on the same search field.

## Acceptance Criteria

- Selecting a provider result includes a valid CSRF token and no longer returns HTTP 403.
- The selected provider schedule opens as expected.
- Invalid or missing CSRF tokens remain rejected.
- Regression coverage exercises searching for and selecting a provider from the Schedule page.

## Checklist

- [x] Searched for similar issues
- [x] Reproduced with synthetic development data
- [x] Verified that no patient data or PHI is included

### PR DESCRIPTION
**Pull request by Ben-Heerema**
## Summary

- render a CSRFGuard token on the Schedule provider-search result page
- copy that token into the dynamically created provider-preference POST before immediate submission
- fail closed when the token is unavailable, and tell the user instead of logging to the console only
- context-encode every provider/group identity rendered by this page, across all four `selectProvider(` call sites and the result table
- fix the null-`Provider` dereference that made every group search (`.`-prefixed) throw before rendering a row
- extend the Schedule Playwright flow to search, select a result, assert the POST token, and fail on HTTP errors

## Testing

- `MAVEN_OPTS="-Xmx1g -Xms256m" mvn -q -Dtest=ScheduleProviderSearchCsrfRegressionTest test` (6 tests)
- `MAVEN_OPTS="-Xmx1500m -Xms256m" mvn -q -DskipTests -Pjspc package` (full JSP compilation; verified the new strings are present in the compiled `receptionistfindprovider_jsp` class, so the JSP was genuinely recompiled)
- `MAVEN_OPTS="-Xmx1g -Xms256m" mvn -q -DskipTests checkstyle:check`
- `bash scripts/lint/check-encoder-null-safety.sh` (PASS, 0 violations)
- `node --check scripts/schedule-links-playwright-checks.js`
- `git diff --check`

The Playwright script is syntax-checked only in this pass; running it needs Tomcat and the dev database up.

## Review round 2

A second review pass found that the guarded `selectProvider()` contract had only been applied to the provider result rows, leaving three defects on the same page.

**MUST — all fixed**

- **Call sites left on the old contract.** The my-group row and the single-result auto-select block still called `selectProvider()` with the old two-argument signature and an unencoded `<%=sp%>`. That value is `MyGroupPrimaryKey.myGroupNo` — an operator-authored `String` — interpolated straight into a JavaScript string literal. The auto-select block is the exact path the new Playwright check exercises, so **the check passed while the unencoded path stayed live**. The my-group anchor also discarded the new fail-closed `return false`.
- **Unencoded provider names.** `spnl` / `spnf` rendered raw into the result-table cells and into the caisi/custom `onClick` handlers — the same `<tr>` rows this PR already edited.
- **Group search was dead.** The `bGrpSearch` branch dereferenced a still-null `Provider` (`p.getLastName()`), so every `.`-prefixed search threw to the error page. The group rows the CSRF fix serves could never render.

**SHOULD — all fixed**

- Missing token failed closed silently (`console.error` only); now surfaced via a new localized `msgMissingSecurityToken` message across all five locale bundles.
- The Playwright fallback click raced a 100 ms wall-clock guess and could throw against an already-navigating popup; the click is now bounded and non-fatal, with the awaited POST as the real signal.
- The new test carried a 4-line copyright stub instead of the standard block from `docs/copyright-header-carlos.md`.

**COULD — one fixed, rest deferred**

- *Fixed:* the browser check hardcoded `CSRF-TOKEN`; it now reads the parameter name off the rendered input, since `org.owasp.csrfguard.TokenName` is configurable.
- *Open:* on the `custom=true` single-result path, `name` is undeclared and resolves to the global `window.name`, writing the popup's window name into the opener's form field. Pre-existing. A first attempt to hoist the variable was **reverted in `ab1d73f0`** because changing it to `''` silently altered what that path sends to the opener — that is a product decision, not an encoding fix, and does not belong in this PR. Recorded with an inline note.
- *Open:* migrate the rest of this legacy search JSP off scriptlets in a separate modernization change.
- *Open:* `.doesNotContain("await searchInput.fill('test');")` pins the absence of a superseded implementation detail; the manual HTTP-error push duplicates what `wirePage` already records.

Closes #3428

## Summary by Sourcery

Secure Schedule provider selection with CSRF token submission, safe identity encoding, and regression coverage for provider and group search flows.

New Features:
- Add CSRF-protected provider selection to the Schedule provider-search flow, including user-visible handling when the token is unavailable.
- Extend the Schedule browser check to select a provider and verify that the resulting POST includes the configured CSRF token.

Bug Fixes:
- Fix group searches so results render without dereferencing a missing provider.
- Prevent unencoded provider and group identities from being inserted into rendered HTML and JavaScript.

Enhancements:
- Ensure all provider-selection call sites use the guarded, fail-closed selection contract.
- Add localized security-token error messages across supported locale bundles.

Tests:
- Add regression coverage for CSRF rendering and submission, encoding, guarded call sites, group searches, and browser-flow token capture.

---
## Summary by cubic
Require a server-rendered CSRF token for schedule provider selection and encode dynamic values. Previously some paths posted without a token and rendered unencoded names; now selection fails closed with a localized alert when the token is missing, and group searches no longer throw.

- JSP: Render a hidden CSRF input via `Owasp.CsrfGuard`, change `selectProvider(p)` to include the token, return false on click, and alert on missing token. Encode the context path, `mygroup_no`, default service type, and provider/group identifiers and names. Fix the group-search null dereference. Keep the custom single-result `name` binding unchanged.
- Browser check: Add a provider last-name search and selection flow. Register the provider POST listener before submit, capture the CSRF parameter name from the search document response, await each parse, and fall back to a DOM read for multi-result lists. Bound waits and report HTTP ≥400. Support `TEST_PROVIDER_LAST_NAME` and `TEST_PROVIDER_SEARCH_ONLY`. Assert the POST includes the CSRF token.
- Tests: Add `ScheduleProviderSearchCsrfRegressionTest` to lock the JSP contract, guarded call sites, encoding, group-search behavior, and browser token-name capture.
- Config: Read the token parameter name from the rendered input instead of assuming `org.owasp.csrfguard.TokenName`. No server changes.

Closes #3428.

<sup>Written for commit 8e5e3d7b9c331bdb7e73889c1c45e2a1f8c0b9a0. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/carlos-emr/carlos/pull/3433?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

File: scripts/schedule-links-playwright-checks.js [hunk 1]
```diff
@@ -166,6 +166,44 @@ async function clickScheduleLink(context, schedulePage, linkSpec) {
...
+  }
+
+  const popupPromise = context.waitForEvent('page', { timeout: 10000 });
+  await searchInput.fill('test');
+  await searchInput.press('Enter');
+  const resultPage = await popupPromise;
+  wirePage(resultPage, 'schedule-provider-search');
+  await resultPage.waitForLoadState('domcontentloaded', { timeout: 30000 });
+
+  const providerLink = resultPage.locator('a[onclick*="selectProvider("]').first();
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply scripts/schedule-links-playwright-checks.js]: Fixed in 6bed420d21. The request listener is now registered before the search is submitted. A single result is allowed to use the JSP auto-submit path; multiple results click the first link; both paths await the same captured POST. The request body and eventual response are then checked. Added a source regression for listener ordering and branch handling.

**Comment by Ben-Heerema** [comment]
Fresh manual re-review after the browser-check timing fix:

- MUST: none
- SHOULD: none
- COULD: none

The listener is installed before navigation, the one-result auto-submit and multi-result click paths cannot double-submit, and missing/no-response cases remain blocking findings. The focused suite passes 3/3; Node syntax and diff checks pass.

@cubic-dev-ai review

File: scripts/schedule-links-playwright-checks.js [hunk 2]
```diff
@@ -166,6 +166,56 @@ async function clickScheduleLink(context, schedulePage, linkSpec) {
...
+    findings.push({ label: 'schedule-provider-search', type: 'missing-search-input' });
+    return;
+  }
+
+  const requestPromise = context.waitForEvent('request', {
+    predicate: (request) => request.method() === 'POST'
+      && /\/provider\/providercontrol(?:\?|$)/.test(request.url()),
+    timeout: 30000,
+  }).catch(() => null);
+  const popupPromise = context.waitForEvent('page', { timeout: 10000 });
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply scripts/schedule-links-playwright-checks.js]: Fixed in aaf59d7bf4. The popup wait now resolves to null on timeout, and that path records a blocking missing-search-popup finding before returning. The focused 3-test suite, Node syntax check, and diff check pass.

**Comment by Ben-Heerema** [comment]
Fresh manual re-review after adding the popup fallback:

- MUST: none
- SHOULD: none
- COULD: none

The timeout path is now a structured blocking finding and cannot dereference a missing page. Focused tests pass 3/3; Node syntax and diff checks pass.

@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
Additional fresh manual review identified and fixed one SHOULD-level harness edge case in `9e57b9c924`: provider-link count is not the JSP total-result count because groups also contribute. The check now uses an already-captured auto-submit request when present and otherwise clicks an available provider link, covering mixed provider/group results without double submission.

Final classification:

- MUST: none
- SHOULD: fixed; none remaining
- COULD: none

Focused tests pass 3/3; Node syntax and diff checks pass.

@cubic-dev-ai review

File: scripts/schedule-links-playwright-checks.js [hunk 3]
```diff
@@ -166,6 +166,63 @@ async function clickScheduleLink(context, schedulePage, linkSpec) {
...
+  const popupPromise = context.waitForEvent('page', { timeout: 10000 }).catch(() => null);
+  await searchInput.fill('test');
+  await searchInput.press('Enter');
+  const resultPage = await popupPromise;
+  if (!resultPage) {
+    findings.push({ label: 'schedule-provider-search', type: 'missing-search-popup' });
+    return;
+  }
+  wirePage(resultPage, 'schedule-provider-search');
+  await resultPage.waitForLoadState('domcontentloaded', { timeout: 30000 });
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply scripts/schedule-links-playwright-checks.js]: Fixed in 24fe93acb0. The load wait now resolves to an explicit boolean; timeout/error records a blocking search-popup-load-failure finding, safely closes an open popup, and returns before any locator access. Focused tests pass 3/3; Node syntax and diff checks pass.

**Comment by Ben-Heerema** [comment]
Fresh manual re-review after the popup-load handling fix:

- MUST: none
- SHOULD: none
- COULD: none

Popup creation, initial load, auto-submit, manual selection, request capture, and response failure paths now all produce deterministic, label-scoped outcomes. Focused tests pass 3/3; Node syntax and diff checks pass.

@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
Final fresh manual re-review of the complete provider-search browser path identified one MUST-level assertion gap and fixed it in `f89b92c5f5`: an auto-submit response may precede page listener wiring, so the captured response is now checked directly for HTTP 4xx/5xx (with duplicate suppression).

Classification after the fix:

- MUST: fixed; none remaining
- SHOULD: none
- COULD: none

The original CSRF 403 class is now explicitly blocking in both auto-submit and manual-click paths. Focused tests pass 3/3; Node syntax and diff checks pass.

@cubic-dev-ai review

File: scripts/schedule-links-playwright-checks.js [hunk 4]
```diff
@@ -166,6 +166,82 @@ async function clickScheduleLink(context, schedulePage, linkSpec) {
...
+    return;
+  }
+
+  const requestPromise = context.waitForEvent('request', {
+    predicate: (request) => request.method() === 'POST'
+      && /\/provider\/providercontrol(?:\?|$)/.test(request.url()),
+    timeout: 30000,
+  }).catch(() => null);
+  const popupPromise = context.waitForEvent('page', { timeout: 10000 }).catch(() => null);
+  await searchInput.fill('test');
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply scripts/schedule-links-playwright-checks.js]: Fixed in 7b08687f3a. The check now defaults to a blank last-name prefix, which lists active providers without assuming a named seed row, and supports TEST_PROVIDER_LAST_NAME when an environment needs a targeted prefix. Regression coverage prevents reintroducing the hard-coded test value.

**Comment by Ben-Heerema** [comment]
Fresh manual re-review after removing seed-name coupling:

- MUST: none
- SHOULD: none
- COULD: none

The default path now selects from existing active providers, while `TEST_PROVIDER_LAST_NAME` allows a targeted environment override. The application behavior and CSRF assertions are unchanged. Focused tests pass 3/3; Node syntax and diff checks pass.

@cubic-dev-ai review

File: src/test/java/io/github/carlos_emr/carlos/provider/ScheduleProviderSearchCsrfRegressionTest.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+                .contains("if (!resultPage)")
+                .contains("type: 'missing-search-popup'")
+                .contains("const resultLoaded = await resultPage.waitForLoadState(")
+                .contains("if (!resultLoaded)")
+                .contains("type: 'search-popup-load-failure'")
+                .contains("let providerRequest = await Promise.race(")
+                .contains("if (!providerRequest && await providerLink.count())")
+                .contains("providerRequest = providerRequest || await requestPromise;")
+                .contains("response.status() >= 400")
+                .contains("finding.type === 'http'");
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/test/java/io/github/carlos_emr/carlos/provider/ScheduleProviderSearchCsrfRegressionTest.java:183]: Fixed in fd13223395. The regression now explicitly guards both the CSRF-TOKEN payload lookup and the missing-csrf-token blocking finding.

File: scripts/schedule-links-playwright-checks.js [hunk 6]
```diff
@@ -166,6 +168,82 @@ async function clickScheduleLink(context, schedulePage, linkSpec) {
...
+    await providerLink.click();
+  }
+
+  providerRequest = providerRequest || await requestPromise;
+  if (!providerRequest) {
+    findings.push({ label: 'schedule-provider-search', type: 'missing-provider-result' });
+    await resultPage.close();
+    return;
+  }
+  const response = await providerRequest.response();
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply scripts/schedule-links-playwright-checks.js]: Fixed in fd13223395. responseWithTimeout bounds request.response() at 30 seconds, clears its timer on either outcome, and the existing provider-post-no-response finding handles timeout or transport failure.

File: src/test/java/io/github/carlos_emr/carlos/provider/ScheduleProviderSearchCsrfRegressionTest.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    @DisplayName("should include server-rendered CSRF token in provider selection POST")
+    void shouldIncludeServerRenderedCsrfToken_inProviderSelectionPost() throws IOException {
+        String jsp = Files.readString(PROVIDER_SEARCH, StandardCharsets.UTF_8);
+
+        assertThat(jsp)
+                .contains("prefix=\"csrf\"")
+                .contains("id=\"providerSelectionCsrfToken\"")
+                .contains("name=\"<csrf:tokenname/>\" value=\"<csrf:tokenvalue/>\"")
+                .contains("document.getElementById('providerSelectionCsrfToken')")
+                .contains("fields[csrfToken.name] = csrfToken.value")
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/test/java/io/github/carlos_emr/carlos/provider/ScheduleProviderSearchCsrfRegressionTest.java:66]: Not changing this as a non-meritable scope expansion. The end-to-end Playwright flow is the behavioral assertion: it submits the rendered page, inspects the actual POST body, and checks its HTTP result. This small source-level test deliberately complements it by pinning JSP compile-time contracts (the CSRF tag rendering and required CARLOS encoding contexts), which cannot be unit-executed without a JSP container. Renaming/refactoring those contracts should update this focused test; replacing the established source-regression pattern with a JSP harness is not warranted in this bug fix.

**Comment by Ben-Heerema** [comment]
Fresh manual re-review after the bounded response wait:

- MUST: none
- SHOULD: none
- COULD: none

The provider request and response phases are independently bounded, timeout is a blocking finding, and the regression now explicitly pins the browser-side missing-CSRF assertion. Cubic’s source-test brittleness suggestion was resolved as non-meritable: behavioral coverage is provided by Playwright, while the source test intentionally guards JSP taglib and encoder compile-time contracts without requiring a container. Focused tests pass 3/3; Node syntax and diff checks pass.

@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

The previous latest-commit run has remained at 94% after both files were reviewed; re-requesting a final verdict on the unchanged head.

**Comment by Ben-Heerema** [comment]
CI note: the sole failing `tests` job is the same unrelated repository-wide `SurefireIncludeCoverageUnitTest` failure on every PR in this series. It reports the pre-existing `DemographicExportAction42ActionRequestMethodTest` on `develop` as unmatched by Surefire; neither that class nor the coverage guard is changed here. This PR’s focused tests and all other completed checks pass.

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
Playwright verification completed on commit d0f2656817 against the local devcontainer Tomcat/MariaDB stack.

The first run reproduced a real harness race: the provider POST returned HTTP 200 with its CSRF token, but the script reported csrf-token-name-unresolved because the popup page event occurs after the initial response begins and every selection navigates the popup. An explicit single-result search also reproduced search-popup-load-failure because auto-submit interrupts the initial DOMContentLoaded wait.

The follow-up now registers the result-response capture at BrowserContext scope before submitting, starts response.text() immediately, and races the provider POST against popup loading. Added TEST_PROVIDER_SEARCH_ONLY=true so this acceptance path can run without unrelated legacy Schedule link-selector findings.

Verified:
- TEST_PROVIDER_SEARCH_ONLY=true node scripts/schedule-links-playwright-checks.js — PASS
- TEST_PROVIDER_SEARCH_ONLY=true TEST_PROVIDER_LAST_NAME=Armstrong node scripts/schedule-links-playwright-checks.js — PASS, single-result auto-submit
- TEST_PROVIDER_SEARCH_ONLY=true TEST_PROVIDER_LAST_NAME=provider node scripts/schedule-links-playwright-checks.js — PASS, multi-result click
- mvn -q -Dtest=ScheduleProviderSearchCsrfRegressionTest test — 6/6
- mvn -q -DskipTests -Pjspc package — PASS
- mvn -q -DskipTests checkstyle:check — PASS
- encoder null-safety lint, node --check, and git diff --check — PASS

**Comment by Ben-Heerema** [comment]
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@yingbull ready for review

**Review by yingbull — APPROVED** [review]
